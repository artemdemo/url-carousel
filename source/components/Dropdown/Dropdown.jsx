import React from 'react';
import classnames from 'classnames';

import './Dropdown.less';

const Dropdown = (props) => {
    const { onChange, value, placeholder, disabled, error, list = [] } = props;

    const itemsList = (() => {
        if (placeholder) {
            return [placeholder].concat(list);
        }
        return list;
    })();

    const dropdownValue = (() => {
        if ((value === '' || value === undefined) && placeholder) {
            return '-1';
        }
        return value;
    })();

    const renderErrorText = (error) => {
        if (typeof error === 'string' && error !== '') {
            return (
                <div className='dropdown-error'>
                    {error}
                </div>
            );
        }
        return null;
    };

    const containerClass = classnames({
        'has-error': Boolean(error),
    });

    return (
        <div className={containerClass}>
            <select
                className='form-control'
                value={dropdownValue}
                onChange={e => onChange && onChange(Number(e.target.value))}
                disabled={disabled}>
                {itemsList.map((item, index) => {
                    return (
                        <option value={placeholder ? index - 1 : index}
                                disabled={placeholder && index - 1 === -1}
                                key={`dropdown__item-${index}`}>{item}</option>
                    );
                })}
            </select>
            {renderErrorText(error)}
        </div>
    );
};

Dropdown.propTypes = {
    onChange: React.PropTypes.func,
    className: React.PropTypes.string,
    list: React.PropTypes.arrayOf(React.PropTypes.string),
    value: React.PropTypes.string,
    placeholder: React.PropTypes.string,
    disabled: React.PropTypes.bool,
    error: React.PropTypes.oneOfType([
        React.PropTypes.bool,
        React.PropTypes.string,
    ]),
};

export default Dropdown;
