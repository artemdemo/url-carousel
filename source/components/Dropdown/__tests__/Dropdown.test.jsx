import React from 'react';
import renderer from 'react-test-renderer';
import Dropdown from '../Dropdown';

const simpleList = ['one', 'two', 'three'];

describe('<Dropdown />', () => {
    it('Render with empty list', () => {
        const tree = renderer.create(
            <Dropdown list={[]} />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('Array of strings', () => {
        const tree = renderer.create(
            <Dropdown list={simpleList} />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('Error', () => {
        const tree = renderer.create(
            <Dropdown list={simpleList} error />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('Error and text', () => {
        const tree = renderer.create(
            <Dropdown list={simpleList} error='Error text' />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('Disabled', () => {
        const tree = renderer.create(
            <Dropdown list={simpleList} disabled />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('With array of strings and placeholder', () => {
        const tree = renderer.create(
            <Dropdown list={simpleList} placeholder='Placeholder string' />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
