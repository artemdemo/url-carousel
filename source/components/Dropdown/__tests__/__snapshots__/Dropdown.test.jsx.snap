// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> Array of strings 1`] = `
<div
  className=""
>
  <select
    className="form-control"
    disabled={undefined}
    onChange={[Function]}
    value={undefined}
  >
    <option
      disabled={undefined}
      value={0}
    >
      one
    </option>
    <option
      disabled={undefined}
      value={1}
    >
      two
    </option>
    <option
      disabled={undefined}
      value={2}
    >
      three
    </option>
  </select>
</div>
`;

exports[`<Dropdown /> Disabled 1`] = `
<div
  className=""
>
  <select
    className="form-control"
    disabled={true}
    onChange={[Function]}
    value={undefined}
  >
    <option
      disabled={undefined}
      value={0}
    >
      one
    </option>
    <option
      disabled={undefined}
      value={1}
    >
      two
    </option>
    <option
      disabled={undefined}
      value={2}
    >
      three
    </option>
  </select>
</div>
`;

exports[`<Dropdown /> Error 1`] = `
<div
  className="has-error"
>
  <select
    className="form-control"
    disabled={undefined}
    onChange={[Function]}
    value={undefined}
  >
    <option
      disabled={undefined}
      value={0}
    >
      one
    </option>
    <option
      disabled={undefined}
      value={1}
    >
      two
    </option>
    <option
      disabled={undefined}
      value={2}
    >
      three
    </option>
  </select>
</div>
`;

exports[`<Dropdown /> Error and text 1`] = `
<div
  className="has-error"
>
  <select
    className="form-control"
    disabled={undefined}
    onChange={[Function]}
    value={undefined}
  >
    <option
      disabled={undefined}
      value={0}
    >
      one
    </option>
    <option
      disabled={undefined}
      value={1}
    >
      two
    </option>
    <option
      disabled={undefined}
      value={2}
    >
      three
    </option>
  </select>
  <div
    className="dropdown-error"
  >
    Error text
  </div>
</div>
`;

exports[`<Dropdown /> Render with empty list 1`] = `
<div
  className=""
>
  <select
    className="form-control"
    disabled={undefined}
    onChange={[Function]}
    value={undefined}
  />
</div>
`;

exports[`<Dropdown /> With array of strings and placeholder 1`] = `
<div
  className=""
>
  <select
    className="form-control"
    disabled={undefined}
    onChange={[Function]}
    value="-1"
  >
    <option
      disabled={true}
      value={-1}
    >
      Placeholder string
    </option>
    <option
      disabled={false}
      value={0}
    >
      one
    </option>
    <option
      disabled={false}
      value={1}
    >
      two
    </option>
    <option
      disabled={false}
      value={2}
    >
      three
    </option>
  </select>
</div>
`;
